crates/sui-graphql-rpc/src/functional_group.rs

```diff
@@ -30,3 +41,228 @@ pub(crate) enum FunctionalGroup {
     /// reference gas price).
     SystemState,
 }
+
+pub(crate) struct FeatureGate;
+
+impl FunctionalGroup {
+    /// Name that the group is referred to by in configuration and responses on the GraphQL API.
+    /// Not a suitable `Display` implementation because it enquotes the representation.
+    fn name(&self) -> String {
+        json::ser::to_string(self).expect("Serializing `FunctionalGroup` cannot fail.")
+    }
+}
+
+impl ExtensionFactory for FeatureGate {
+    fn create(&self) -> Arc<dyn Extension> {
+        Arc::new(FeatureGate)
+    }
+}
+
+#[async_trait]
+impl Extension for FeatureGate {
+    async fn resolve(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        info: ResolveInfo<'_>,
+        next: NextResolve<'_>,
+    ) -> ServerResult<Option<Value>> {
+        let ResolveInfo {
+            parent_type,
+            name,
+            is_for_introspection,
+            ..
+        } = &info;
+
+        let ServiceConfig {
+            disabled_features, ..
+        } = ctx.data().map_err(|_| {
+            graphql_error(
+                code::INTERNAL_SERVER_ERROR,
+                "Unable to fetch service configuration",
+            )
+        })?;
+
+        // TODO: Is there a way to set `is_visible` on `MetaField` and `MetaType` in a generic way
+        // after building the schema? (to a function which reads the `ServiceConfig` from the
+        // `Context`). This is (probably) required to hide disabled types and interfaces in the
+        // schema.
+
+        if let Some(group) = functional_group(parent_type, name) {
+            if disabled_features.contains(&group) {
+                return if *is_for_introspection {
+                    Ok(None)
+                } else {
+                    Err(ServerError::new(
+                        format!(
+                            "Cannot query field \"{name}\" on type \"{parent_type}\". \
+                             Feature {} is disabled.",
+                            group.name(),
+                        ),
+                        // TODO: Fork `async-graphl` to add field position information to
```

Comment:
Instead of forking how about contributing back this feature?

Reply:
Yes indeed, that's the intention, just don't want to block our own progress on waiting for that to come back via upstream!  So, I would fork, make the change, and in parallel we can upstream it and also use it here.